ssl/ech_local.h

```diff
@@ -0,0 +1,588 @@
+#  define OSSL_ECH_SAME_EXT_DONE 1 /* proceed with same value in inner/outer */
+#  define OSSL_ECH_SAME_EXT_CONTINUE 2 /* generate a new value for outer CH */
+
+/*
+ * During extension construction (in extensions_clnt.c
+ * and surprisingly also in extensions.c), we need to
+ * handle inner/outer CH cloning - ech_same_ext will
+ * (depending on ech.c compile time handling options
+ * from the ech_ext_handling table) copy the value
+ * from CH.inner to CH.outer or else processing
+ * will continue, for a 2nd call, likely generating a
+ * fresh value for the outer CH. (The fresh value
+ * could well be the same as in the inner.)
+ * This macro should be called in each _ctos_ function
+ * that doesn't explicitly have special ECH handling.
+ *
+ * Note that the placement of this macro needs a bit
+ * of thought - it has to go after declarations (to
+ * keep the ansi-c compile happy) and also after any
+ * checks that result in the extension not being sent
+ * but before any relevant state changes that would
+ * affect a possible 2nd call to the constructor.
+ * Luckily, that's usually not to hard, but it's
+ * not mechanical.
+ */
+#  define ECH_IOSAME(s, pkt) \
+    if (s->ext.ech.cfgs != NULL && s->ext.ech.grease == 0) { \
+        int __rv = ech_same_ext(s, pkt); \
+        \
+        if (__rv == OSSL_ECH_SAME_EXT_ERR) \
+            return EXT_RETURN_FAIL; \
+        if (__rv == OSSL_ECH_SAME_EXT_DONE) \
+            return EXT_RETURN_SENT; \
+        /* otherwise continue as normal */ \
+    }
+
+/*
+ * Reminder of what goes in DNS for ECH RFC XXXX
+ *
+ *     opaque HpkePublicKey<1..2^16-1>;
+ *     uint16 HpkeKemId;  // Defined in I-D.irtf-cfrg-hpke
+ *     uint16 HpkeKdfId;  // Defined in I-D.irtf-cfrg-hpke
+ *     uint16 HpkeAeadId; // Defined in I-D.irtf-cfrg-hpke
+ *     struct {
+ *         HpkeKdfId kdf_id;
+ *         HpkeAeadId aead_id;
+ *     } HpkeSymmetricCipherSuite;
+ *     struct {
+ *         uint8 config_id;
+ *         HpkeKemId kem_id;
+ *         HpkePublicKey public_key;
+ *         HpkeSymmetricCipherSuite cipher_suites<4..2^16-4>;
+ *     } HpkeKeyConfig;
+ *     struct {
+ *         HpkeKeyConfig key_config;
+ *         uint8 maximum_name_length;
+ *         opaque public_name<1..255>;
+ *         Extension extensions<0..2^16-1>;
+ *     } ECHConfigContents;
+ *     struct {
+ *         uint16 version;
+ *         uint16 length;
+ *         select (ECHConfig.version) {
+ *           case 0xfe0d: ECHConfigContents contents;
+ *         }
+ *     } ECHConfig;
+ *     ECHConfig ECHConfigList<1..2^16-1>;
+ */
+typedef unsigned char ech_ciphersuite_t[OSSL_ECH_CIPHER_LEN];
+
+typedef struct ech_config_st {
+    unsigned int version; /* 0xff0d for draft-13 */
+    unsigned int public_name_len;
+    unsigned char *public_name;
+    unsigned int kem_id;
+    unsigned int pub_len;
+    unsigned char *pub;
+    unsigned int nsuites;
+    ech_ciphersuite_t *ciphersuites;
+    unsigned int maximum_name_length;
```

Comment:
'uint8'  ( or `uint8_t` ?)  is already used in current codebase.

Reply:
I think I had that at one stage but changed back for some reason (maybe one of the ports run in the workflows) but I don't recall. I've added a TODO to look again at the types used in ECHConfig and ECHConfigList ('cause as you say other structs do use uint16_t etc now).

Reply:
So I checked this out and I think those types are best left as-is. As a test I changed the ``version`` field to be a ``uint16_t`` but then that generates compile warnings such as the one below due to the PACKET APis dealing with unsigned ints. I'd argue that making such changes would therefore just add lots of casts that could be more error-prone and no real benefit, other than slightly smaller storage for structs which are not expected to be big in real use-cases anyway.

```
ssl/ech.c: In function 'ECHConfigList_from_binary':
ssl/ech.c:801:40: warning: passing argument 2 of 'PACKET_get_net_2' from incompatible pointer type [-Wincompatible-pointer-types]
  801 |             || !PACKET_get_net_2(&pkt, &ec->version)
      |                                        ^~~~~~~~~~~~
      |                                        |
      |                                        uint16_t * {aka short unsigned int *}
In file included from ssl/ssl_local.h:32,
                 from ssl/ech.c:11:
include/internal/packet.h:150:75: note: expected 'unsigned int *' but argument is of type 'uint16_t *' {aka 'short unsigned int *'}
  150 | __owur static ossl_inline int PACKET_get_net_2(PACKET *pkt, unsigned int *data)
```

Reply:
I replaced the TODO I'd put in for this with a comment reflecting the above. Will push that next time.